From 7f906e3544e091aeaff91d86568cfcf7fcda6ab9 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Thu, 23 May 2013 14:36:28 -0700
Subject: [PATCH 002/974] MIPS: Octeon: Fast access thread pointer

If CONFIG_FAST_ACCESS_TO_THREAD_POINTER is enabled, then deploy the
FAST_ACCESS_THREAD_OFFSET and the FAST_ACCESS_THREAD_REGISTER to
bounce through the reserved hardware cache area for speed.

Signed-off-by: David Daney <david.daney@cavium.com>
[Original patch taken from Cavium SDK 3.1.2-568]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 arch/mips/include/asm/mipsregs.h | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/arch/mips/include/asm/mipsregs.h b/arch/mips/include/asm/mipsregs.h
index 3d0074e..569cbbd 100644
--- a/arch/mips/include/asm/mipsregs.h
+++ b/arch/mips/include/asm/mipsregs.h
@@ -622,6 +622,21 @@
 #define MIPS_FPIR_L		(_ULCAST_(1) << 21)
 #define MIPS_FPIR_F64		(_ULCAST_(1) << 22)
 
+/*
+ * These defines are used on Octeon to implement fast access to the
+ * thread pointer from userspace. Octeon uses a 64bit location in
+ * CVMSEG to store the thread pointer for quick access.
+ *
+ * TLB refill uses location -8, fast access is -16 (both from the top
+ * of the area.
+ */
+#ifdef CONFIG_FAST_ACCESS_TO_THREAD_POINTER
+#define FAST_ACCESS_THREAD_OFFSET			\
+	(CONFIG_CAVIUM_OCTEON_CVMSEG_SIZE*128-16-32768)
+#define FAST_ACCESS_THREAD_REGISTER			\
+	(*(unsigned long *)(FAST_ACCESS_THREAD_OFFSET))
+#endif
+
 #ifndef __ASSEMBLY__
 
 /*
-- 
2.6.2

